plugins/router/f5/plugin.go

```diff
@@ -0,0 +1,529 @@
+// the given endpoints.
+func (p *F5Plugin) updatePool(poolname string, endpoints *kapi.Endpoints) error {
+	members, err := p.F5Client.GetPoolMembers(poolname)
+	if err != nil {
+		glog.V(4).Infof("F5Client.GetPoolMembers failed: %v", err)
+		return err
+	}
+
+	// We need to keep track of which endpoints already existed in F5 in order
+	// to delete any that no longer exist in the updated set of endpoints.
+	//
+	// It would be really nifty if F5 would just let us PUT the new list to
+	// the pool members resource instead of making a bunch of POST and DELETE
+	// requests, bu-u-ut... it doesn't.
+	needToDelete := map[string]bool{}
+	for member, _ := range members {
+		if members[member] {
+			needToDelete[member] = true
+		}
+	}
+
+	// Note that OpenShift issues many spurious notifications for updates when
+	// the endpoints set is actually the same, so we may ultimately end up
+	// adding and deleting 0 endpoints.
+
+	for _, subset := range endpoints.Subsets {
+		for _, addr := range subset.Addresses {
+			for _, port := range subset.Ports {
+				dest := fmt.Sprintf("%s:%d", addr.IP, port.Port)
+				exists := needToDelete[dest]
+				needToDelete[dest] = false
+				if exists {
+					glog.V(4).Infof("  Skipping %s because it already exists.", dest)
+				} else {
+					glog.V(4).Infof("  Adding %s...", dest)
+					err = p.F5Client.AddPoolMember(poolname, dest)
+					if err != nil {
+						glog.V(4).Infof("  Error adding endpoint %s to pool %s: %v",
+							dest, poolname, err)
+					}
+				}
+			}
+		}
+	}
+
+	for member, _ := range needToDelete {
+		if needToDelete[member] {
+			glog.V(4).Infof("  Deleting %s...", member)
+			err = p.F5Client.DeletePoolMember(poolname, member)
+			if err != nil {
+				glog.V(4).Infof("  Error deleting endpoint %s from pool %s: %v",
+					member, poolname, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// deletePool delete the named pool from F5 BIG-IP.
+func (p *F5Plugin) deletePool(poolname string) error {
+	poolExists, err := p.F5Client.PoolExists(poolname)
+	if err != nil {
+		glog.V(4).Infof("F5Client.PoolExists failed: %v", err)
+		return err
+	}
+
+	if poolExists {
+		err = p.F5Client.DeletePool(poolname)
+		if err != nil {
+			glog.V(4).Infof("Error deleting pool %s: %v", poolname, err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+// deletePoolIfEmpty deletes the named pool from F5 BIG-IP if, and only if, it
+// has no members.
+func (p *F5Plugin) deletePoolIfEmpty(poolname string) error {
+	poolExists, err := p.F5Client.PoolExists(poolname)
+	if err != nil {
+		glog.V(4).Infof("F5Client.PoolExists failed: %v", err)
+		return err
+	}
+
+	if poolExists {
+		members, err := p.F5Client.GetPoolMembers(poolname)
+		if err != nil {
+			glog.V(4).Infof("F5Client.GetPoolMembers failed: %v", err)
+			return err
+		}
+
+		// We only delete the pool if the pool is empty, which it may not be
+		// if a service has been added and has not (yet) been deleted.
+		if len(members) <= 0 {
```

Comment:
Yeah, I like the lvalue-on-RHS for comparisons technique and broader conditional too.  I actually observed the latter in the template router, which is why I went with it here, but if it goes against the wider convention in the OpenShift codebase, I can change this comparison.